src/xrGame/space_restriction_shape.cpp

```diff
@@ -79,10 +80,45 @@ void CSpaceRestrictionShape::fill_shape(const CCF_Shape::shape_def& shape)
     }
     default: NODEFAULT;
     }
-    ai().level_graph().iterate_vertices(start, dest, CBorderMergePredicate(this));
+
+    CLevelGraph& graph = ai().level_graph();
+
+    std::mutex mergeMutex;
+
+    auto [begin, end] = graph.get_range(start, dest);
+    xr_parallel_for(TaskRange(begin, end), [this, &mergeMutex, &graph] (const auto& range)
+    {
+        xr_vector<u32> m_border_chunk;
+        m_border_chunk.reserve(range.size());
+        for (const auto& vertex : range)
+        {
+            if (inside(graph.vertex_id(&vertex), true) &&
+                !inside(graph.vertex_id(&vertex), false))
+                m_border_chunk.push_back(graph.vertex_id(&vertex));
+        }
+        std::lock_guard lock(mergeMutex);
+        if (m_border.capacity() < m_border.size() + m_border_chunk.size())
+            m_border.reserve(m_border.size() + m_border_chunk.size());
+        for (auto x : m_border_chunk)
+            m_border.push_back(x);
```

Comment:
I did a little research on this, and it seemed to me that the body `CBorderMergePredicate::operator(...)` is a bigger problem than merging a certain number of chunks under lock
![image](https://github.com/OpenXRay/xray-16/assets/68744350/25600259-56e2-4c77-bd7b-c8d7abddcc90)

Reply:
Ok, it might be irrelevant since I accidentally used `UI Freeze` option.
I'm sorry for the misinformation. Apparently, this flame graph represents only those samples that were sampled when the NPC's spawning was lagging.

> The UI Freeze event indicates time intervals where the application was unable to respond to user input. More specifically, these are time intervals where window messages were not pumped for more than 200 ms or processing of a particular message took more than 200 ms.

Reply:
@olefirenque, what profiler did you use?

Reply:
To me this strongly looks like [tracy](https://github.com/wolfpld/tracy) but I might be wrong there